src/utilities.js

```diff
@@ -437,3 +438,43 @@ exports.promisifyServerListen = (server) => {
     };
 };
 
+const gzipP = Promise.promisify(gzip);
+/**
+ * Prepares an array of data for dispatch over network connection by serialization and compression into
+ * one or multiple chunks, limited in size by the maxChunkSizeBytes parameter.
+ *
+ * @param {Array} array of JSON serializable values
+ * @param {Number} maxChunkSizeBytes
+ *
+ * @return {Array}
+ */
+exports.stringifyGzipChunkArray = function stringifyGzipChunkArray(array, maxChunkSizeBytes) { // using "function" to be able to self-reference
+    if (!array.length) return [];
+    const serialized = JSON.stringify(array);
+    return gzipP(serialized).then((zipped) => {
+        const byteSize = Buffer.byteLength(zipped);
+        // return fast if size is ok
+        if (byteSize < maxChunkSizeBytes) return [zipped];
+        // now we know that it isn't, first check if we can do anything about it
+        if (array.length === 1) throw new Error(`Unable to serialize: Chunk too large! Item is ${byteSize} bytes. Limit is ${maxChunkSizeBytes} bytes.`); // eslint-disable-line
+        if (array.length > 1 && Math.ceil(byteSize / array.length) > maxChunkSizeBytes) throw new Error(`Unable to serialize: Data too large! Array cannot be split into chunks < ${maxChunkSizeBytes} bytes.`); // eslint-disable-line
+        // theoretically possible to chunk now, let's assume that all items are roughly equal in size
+        const numOfChunks = Math.ceil(byteSize / maxChunkSizeBytes);
```

Comment:
I would chunk it to lets say half (constant) of maxChunkSizeBytes. Because of this way you very likely excess maxChunkSizeBytes again.

Reply:
Or even third. If max payload is 9MB then making requests for 3MB is not much overhead.